Link fixes + SessionSigs Updates (#311)
LIT-Protocol/docs: docs/sdk/authentication/session-sigs/get-session-sigs.md

---

import FeedbackComponent from "@site/src/pages/feedback.md";
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # Generating SessionSigs: `getSessionSigs`

Reviewer: ```suggestion
# Generating SessionSigs Using an Auth Sig
```

Author: Did this, but `AuthSig` or `Auth Sig`?

Author: Will make this Auth Sig, I think people will understand

This guide covers the `getSessionSigs` function from the Lit SDK. For an overview of what session signatures are and how they are to be used, please go [here](./intro).

Using the `getSessionSigs` function, you can specify the capabilities of your current session on the Lit network.

The function uses the session keypair generated by the `LitNodeClient` and invokes a callback function, `authNeededCallback`, to generate an `AuthSig` scoped to specific Lit capabilities and sign it using the session keypair to create the `SessionSigs`.

Using the `getSessionSigs` function, you can specify the capabilities of your current session on the Lit network. This function is the simplest way to get session signatures, at the minimum only requiring an EOA Ethereum private key and the `LitNodeClient`. It will enable specific capabilities for your session keypair using the resources you specify in the `AuthSig`.

## Prerequisites

Before continuing this guide, you should have an understanding of:
- [Session Signatures](./intro)
- [Lit Resources and Abilities](./resources-and-abilities.md)

> ## Parameters and Returns

Reviewer: ```suggestion
## Parameters and Returns Values
```

To see the parameters and return of `getSessionSigs`, please visit our [API Docs](https://v6-api-doc-lit-js-sdk.vercel.app/classes/lit_node_client_src.LitNodeClientNodeJs.html#getSessionSigs).

## Example Implementation

The full code implementation can be found [here](https://github.com/LIT-Protocol/developer-guides-code/tree/master/session-signatures/getSessionSigs). 

### Installing the Required Dependencies
<Tabs
defaultValue="npm"
values={[
{label: 'npm', value: 'npm'},
{label: 'yarn', value: 'yarn'},
]}>
<TabItem value="npm">

```bash
npm install \
@lit-protocol/auth-helpers \
@lit-protocol/constants \
@lit-protocol/lit-node-client \
ethers@v5
```

</TabItem>

<TabItem value="yarn">

```bash
yarn add \
@lit-protocol/auth-helpers \
@lit-protocol/constants \
@lit-protocol/lit-node-client \
ethers@v5
```

</TabItem>
</Tabs>

> The `node-localstorage` dependency is only required when executing code outside a browser environment. The SDK will use the native browser storage when in a browser environment.

Reviewer: ```suggestion
The `node-localstorage` dependency is only required when executing code outside a browser environment. The SDK will use the native browser storage when in a browser environment. You can learn more about this [here](../intro#storing-sessionsigs).
```

```ts
import { LitNodeClient } from "@lit-protocol/lit-node-client";
import { LitNetwork } from "@lit-protocol/constants";

let litNodeClient: LitNodeClient;

litNodeClient = new LitNodeClient({
      litNetwork: LitNetwork.DatilTest,
      debug: false,
    });
await litNodeClient.connect();
```

### Generating Session Signatures
In this example, we're granting the capability to request to decrypt any data that we may be authorized to decrypt (i.e. we satisfy the Access Control Conditions the data was encrypted with). We could, however, specify the [LitAccessControlConditionResource](https://v6-api-doc-lit-js-sdk.vercel.app/classes/auth_helpers_src.LitAccessControlConditionResource.html) for specific encrypted data we're permitting the decryption capability for. In real-world applications, it's more common and secure to limit access to specific Lit resources instead of specifying the wildcard (`"*"`) identifier.

To get the Lit resource identifier for other resources, you can use the other methods included in [@lit-protocol/auth-helpers](https://v6-api-doc-lit-js-sdk.vercel.app/modules/auth_helpers_src.html) package.

> If you would like to use this function on the `datil` or `datil-test` networks, a `capacityDelegationAuthSig` is required. An example of how to generate one can be found in the full code example.

Reviewer: Once #319 is merged, we should like to the Delegating a Credit page

---

import FeedbackComponent from "@site/src/pages/feedback.md";
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Generating SessionSigs: `getSessionSigs`

This guide covers the `getSessionSigs` function from the Lit SDK. For an overview of what session signatures are and how they are to be used, please go [here](./intro).

> The function uses the session keypair generated by the `LitNodeClient` to invoke a callback function that signs the generated session key, creating an `AuthSig` that is scoped to specific capabilities.

Reviewer: ```suggestion
The function uses the session keypair generated by the `LitNodeClient` to invoke a callback function that signs the generated session key, creating an `AuthSig` that is scoped to specific Lit capabilities.
```

> Using the `getSessionSigs` function, you can specify the abilities of your current session on the Lit network. This function is the simplest way to get session signatures, at the minimum only requiring an `ethers.Wallet` and the `LitNodeClient`. It will enable your `ethers.Wallet` to perform abilities to the resources you specify in the `AuthSig`.

Reviewer: ```suggestion
Using the `getSessionSigs` function, you can specify the Lit capabilities of your current session on the Lit network. This function is the simplest way to get session signatures, at the minimum only requiring an `ethers.Wallet` and the `LitNodeClient`. It will enable your `ethers.Wallet` to perform abilities to the resources you specify in the `AuthSig`.
```

## Prerequisites

Before continuing this guide, you should have an understanding of:
- [Session Signatures](./intro)
- [Lit Resources and Abilities](./resources-and-abilities.md)

## `getSessionSigs`' Interface
[Source code](https://github.com/LIT-Protocol/js-sdk/blob/master/packages/types/src/lib/interfaces.ts#L1057-L1131)

```ts
interface CommonGetSessionSigsProps {
  pkpPublicKey?: string;
  expiration?: any;
  chain?: Chain;
  resourceAbilityRequests: LitResourceAbilityRequest[];
  sessionCapabilityObject?: ISessionCapabilityObject;
  switchChain?: boolean;
  sessionKey?: SessionKeyPair;
  capabilityAuthSigs?: AuthSig[];
}

interface GetSessionSigsProps
  extends CommonGetSessionSigsProps,
    LitActionSdkParams {
  authNeededCallback: AuthCallback;
}

getSessionSigs = async (
    params: GetSessionSigsProps
  ): Promise<SessionSigsMap>

interface AuthSig {
  sig: any;
  derivedVia: string;
  signedMessage: string;
  address: string;
}

type SessionSigsMap = Record<string, AuthSig>;

> ```

Reviewer: From Daryl's feedback, we shouldn't include links to the source code or have the source code included in the docs because this is hard to maintain whenever there's a change to the interface within the SDK. So, this section should be something like:

```
## `getSessionSigs` Interface

For details on this methods interface, please go [here](https://v6-api-doc-lit-js-sdk.vercel.app/classes/lit_node_client_src.LitNodeClientNodeJs.html#getSessionSigs).
```

Also open to other ideas of how this should look like if you have them

Reviewer: And this would imply we remove the `Parameters` and `Return Value` sections, as this info should be in the SDK API docs. For this I think we should review the SDK API docs and if the information isn't better than what you have here, we should make a PR into the SDK to improve the docs

Author: I guess if Daryl doesn't recommend any source code (snippits or links) in the docs then we can only reference the API docs. I agree with just linking the API docs then.

I based my `Parameters` and `Return Value` sections off of the `Wrapped Keys` guides, with most of the information being directly from the SDK (so via the API docs). I think explaining parameters/returns in the docs is worthwhile even if the information is in the API docs as well, looks better and makes the concepts more understandable as well IMO. Could we get a third opinion on this?

Reviewer: > I think explaining parameters/returns in the docs is worthwhile even if the information is in the API docs as well, looks better and makes the concepts more understandable as well IMO

I agree, but the problem Daryl raised is we now have to manage the information in both the docs and the SDK API docs, instead of just directing users to the API docs which will (in theory) always be kept up to date since the docs live next to the code

If the SDK API docs aren't covering what each parameter/return value is sufficiently, then we should improve those docs, not write them in two places

Author: Alright, I removed the sections from the PR. I don't think the SDK API Docs are very good, so I'll make a PR with better explanations

### `signedMessage`
The payload signed by the session key pair. This is the signed `AuthSig`, which includes the fields:

- #### `sessionKey`
   The session key pair public key.
- #### `resourceAbilityRequests`
   An array of resource abilities that you have requested for the session.
- #### `capabilities` <- Needs a fact check
   An array of one or more `AuthSig`. The `capabilities` authorize the `AuthSig` address(es) to utilize the resources specified. in the `resourceAbilityRequests`
- #### `issuedAt`
   The time the session signature was issued.
- #### `expiration`
   The time the session signature becomes invalid.
- #### `nodeAddress`
   The specific URL the session signature is meant for.

### `address`
The session key pair public key.

### `algo`
The signing algorithm used to generate the session signature.

## Example Implementation

Now that we know what the `getSessionSigs` function does, it's parameters, and it's return values, let's dig into a complete implementation.

The full code implementation can be found [here](https://github.com/LIT-Protocol/developer-guides-code/tree/master/session-signatures/getSessionSigs). 

> ### Installing the Required Dependencies

Reviewer: We should add a note to this section that the `node-localstorage` dependency is only required when executing outside of a browser environment since the SDK will use the native browser storage

<Tabs
defaultValue="npm"
values={[
{label: 'npm', value: 'npm'},
{label: 'yarn', value: 'yarn'},
]}>
<TabItem value="npm">

```bash
npm install \
@lit-protocol/auth-helpers \
@lit-protocol/constants \
@lit-protocol/lit-node-client \
node-localstorage \
ethers@v5
```

</TabItem>

<TabItem value="yarn">

```bash
yarn add \
@lit-protocol/auth-helpers \
@lit-protocol/constants \
@lit-protocol/lit-node-client \
node-localstorage \
ethers@v5
```

</TabItem>
</Tabs>

### Instantiating an Ethers Signer
The `ETHEREUM_PRIVATE_KEY` environment variable is required.
```ts
import { LIT_RPC } from "@lit-protocol/constants";
import * as ethers from "ethers";

const ethersSigner = new ethers.Wallet(

>   ETHEREUM_PRIVATE_KEY,

Reviewer: ```suggestion
  process.env.ETHEREUM_PRIVATE_KEY,
```

We should always prefix ENVs with `process.env.`

  new ethers.providers.JsonRpcProvider(LIT_RPC.CHRONICLE_YELLOWSTONE)
);
```

### Instantiating a `LitNodeClient`

> Here we are instantiating an instance of `LitNodeClient` and connecting it to the `datil-dev` Lit network. To simulate a Node.js environment, we use `LocalStorage` from the `node-localstorage` module to store our session keys and metadata.

Reviewer: ```suggestion
Here we are instantiating an instance of `LitNodeClient` and connecting it to the `datil-dev` Lit network. Because we're executing within a Node.js environment, we use `LocalStorage` from the `node-localstorage` module to store our session keys and metadata.
```

```ts
import { LitNodeClient } from "@lit-protocol/lit-node-client";
import { LitNetwork } from "@lit-protocol/constants";
import { LocalStorage } from "node-localstorage";

let litNodeClient: LitNodeClient;

litNodeClient = new LitNodeClient({
      litNetwork: LitNetwork.DatilDev,
      debug: false,
      storageProvider: {
        provider: new LocalStorage("./lit_storage.db"),
      },
    });
await litNodeClient.connect();
```

### Generating Session Signatures

> The session signatures generated in this example provide unrestricted access to all resources. However, in real-world applications, it's more common and secure to limit access. Typically, you would use the specific hash of the encrypted access control condition resource key, granting access only to particular resources rather than everything. To this resource, we then give it the ability to decrypt an encryption access control condition. 

Reviewer: ```suggestion
In this example, we're granting the capability to request to decrypt any data that we may be authorized to decrypt (i.e. we satisfy the Access Control Conditions the data was encrypted with). We could, however, specify the [LitAccessControlConditionResource](https://v6-api-doc-lit-js-sdk.vercel.app/classes/auth_helpers_src.LitAccessControlConditionResource.html) for specific encrypted data we're permitting the decryption capability for. In real-world applications, it's more common and secure to limit access to specific Lit resources instead of specifying the wildcard (`"*"`) identifier.

To get the Lit resource identifier for other resources, you can use the other methods included in [@lit-protocol/auth-helpers](https://v6-api-doc-lit-js-sdk.vercel.app/modules/auth_helpers_src.html) package.
```

```ts
import {
  LitAbility,
  LitAccessControlConditionResource,
  createSiweMessage,
  generateAuthSig,
} from "@lit-protocol/auth-helpers";

const sessionSignatures = await litNodeClient.getSessionSigs({
  chain: "ethereum",
  expiration: new Date(Date.now() + 1000 * 60 * 10 ).toISOString(), // 10 minutes
  resourceAbilityRequests: [
    {
      resource: new LitAccessControlConditionResource("*"),
      ability: LitAbility.AccessControlConditionDecryption,
    },
  ],
  authNeededCallback: async ({
    uri,
    expiration,
    resourceAbilityRequests,
  }) => {
    const toSign = await createSiweMessage({
      uri,
      expiration,
      resources: resourceAbilityRequests,

>       walletAddress: await ethersSigner.getAddress(),

Reviewer: ```suggestion
      walletAddress: wallet.address,
```

This should work

Author: Wallet isn't defined in the code example, ethersSigner.address seems to be alright

      nonce: await litNodeClient.getLatestBlockhash(),
      litNodeClient,
    });

    return await generateAuthSig({
      signer: ethersSigner,
      toSign,
    });
  },
});
```

> **Note:** The nonce should be the latest Ethereum blockhash returned by the nodes during the handshake.

Reviewer: ```suggestion
:::note
The nonce should be the latest Ethereum blockhash returned by the nodes during the handshake.
:::
```

> ## Resources You Can Request

Reviewer: I feel like this section can be removed, and we should cover it in the `Session Signatures` overview page, or in the specific page `Lit Resources and Abilities`

I'd actually like to redo the `Lit Resources and Abilities` and `Capability Objects` pages, but that's out of scope for this PR

Author: I'll add this to the overview

Author: Will say this in the overview. I do think the `Lit Resources and Abilities` page could be redone, maybe I get to it after wrapping up session sigs?

Reviewer: I've created [a ticket](https://linear.app/litprotocol/issue/DREL-69/refactor-the-lit-resources-and-abilities-docs-page) for it, but we should discuss priority in the next stand up

> You can pass an array of `resourceAbilityRequests` to the above functions, which will be presented to the user in the SIWE message - read more [here](resources-and-abilities) about Lit resources and abilities. The resources and abilities requested by the session key must be narrower or equal to the capabilities granted to it per the session capability object specified in the inner `AuthSig`. 

Reviewer: ```suggestion
You can pass an array of `resourceAbilityRequests` to the `getSessionSigs` function, which will be presented to the user in the SIWE message - read more [here](resources-and-abilities) about Lit resources and abilities. The resources and abilities requested by the session key must be narrower or equal to the capabilities granted to it per the session capability object specified in the inner `AuthSig`. 
```

When session capability objects are omitted from functions generating session signatures, the SDK will generate a session capability object with **wildcard permissions against all of the resources in that category by default**, i.e. ability to perform operations against all access control conditions. Read more [here](capability-objects) about how to create custom session capability objects.

> ## Clearing Local Storage

Reviewer: Should add an explanation of why you'd want to do this. Although I'm not sure why you'd want to outside of debugging

Author: Will say this in the overview

If you want to clear the session key stored in the browser local storage, you can call the [`disconnectWeb3` method](https://js-sdk.litprotocol.com/functions/auth_browser_src.ethConnect.disconnectWeb3.html).

## Summary
The full code implementation can be found [here](https://github.com/LIT-Protocol/developer-guides-code/tree/master/session-signatures/getSessionSigs). 

> After executing the example implementation above, you will have generated and stored your session public and secret keys, and the `AuthSig` signed for the session signatures. 

Reviewer: ```suggestion
After executing the example implementation above, you will have generated and stored session signatures that allow you to request to decrypt data that you satisfy the Access Control Conditions for.

You can use these session signatures in a decryption request [like so](../../../access-control/quick-start#decryption).
```

Author: What about:

After executing the example implementation above, you will have generated and stored session signatures that allow requests to decrypt data that match your Access Control Conditions.

Reviewer: ```
After executing the example implementation above, you will have generated and stored session signatures that allow requests to decrypt data that you satisfy the Access Control Conditions for.
```

Were not matching the control conditions, they're a set of constraints that we're satisfying

# Generating SessionSigs

The Lit SDK has three functions that explicitly generate session signatures: `getSessionSigs()`, `getPkpSessionSigs()`, and `getLitActionSessionSigs()`. These functions differ in their parameters and uses, but will all return session signatures. 

These functions use the session keypair generated by the `LitNodeClient` to invoke a callback function that signs the generated session key, creating an `AuthSig` that is scoped to specific capabilities.

> In order to generate `SessionSigs` on both the `Habanero` and `Manzano` networks, you will need to include `capabilityAuthSigs`. This is because transactions on these networks require capacity credits. `capabilityAuthSigs` are **not** required on the `datil-dev` network.

Reviewer 2: we might also want to reference the 'upcoming' datil-test and datil-prod networks as well here

      },
    ],
    authNeededCallback,
  });

  return sessionSigs;
};
```

**Note:** The nonce should be the latest Ethereum blockhash returned by the nodes during the handshake.

## getPkpSessionSigs()

> This is an implementation of `getPkpSessionSigs()`. Noticably different from `getSessionSigs()`, it lacks the `authNeededCallback` parameter, instead creating the callback and signing the session key under the hood with the `signSessionKey()` function. [API Docs](https://v6-api-doc-lit-js-sdk.vercel.app/classes/lit_node_client_src.LitNodeClientNodeJs.html#getLitActionSessionSigs)

Reviewer 2: can we specify why someone would want to use one of these functions over the other? Like hint at recommended usage ('we recommend using this function if you are...') perhaps in the overview at the beginning of this section?

Author: @a1ttech Changed this, would appreciate confirmation on the accuracy of my comments 😅:

We recommend using `getSessionSigs()` to specify the abilities of your current session on the Lit Network, using only an `ethers.Wallet` and the `LitNodeClient`. This method is arguably the simplest, enabling your `ethers.Wallet` to perform the specified abilities on the Lit Network. `getPkpSessionSigs()` can be used if your implementation of the Lit SDK involves a PKP and you wish to designate resources to it. Similarly, `getLitActionSessionSigs()` can designate specific abilities and resources to Lit Actions, potentially allowing them to execute code on the Lit Network.
